lua/oil/adapters/trash/windows.lua

```diff
@@ -1,20 +1,339 @@
+    if src_adapter.name == "files" then
+      local _, path = util.parse_url(action.src_url)
+      assert(path)
+      local short_path = files.to_short_os_path(path, action.entry_type)
+      return string.format("  COPY %s -> TRASH", short_path)
+    elseif dest_adapter.name == "files" then
+      local _, path = util.parse_url(action.dest_url)
+      assert(path)
+      local short_path = files.to_short_os_path(path, action.entry_type)
+      return string.format("RESTORE %s", short_path)
+    else
+      error("Must be copying files into or out of trash")
+    end
+  else
+    error(string.format("Bad action type '%s'", action.type))
+  end
+end
+
+---@param action oil.Action
+---@param cb fun(err: nil|string)
+M.perform_action = function(action, cb)
+  if action.type == "create" then
+    -- TODO: do nothing?
+  elseif action.type == "delete" then
+    local entry = assert(cache.get_entry_by_url(action.url))
+    local meta = entry[FIELD_META] --[[@as {stat: uv_fs_t, trash_info: oil.TrashInfo, display_name: string}]]
+    local trash_info = meta.trash_info
+
+    local jid = vim.fn.jobstart({
+      "powershell",
+      "-NoProfile",
+      "-ExecutionPolicy",
+      "Bypass",
+      "-Command",
+      ([[
+$path = Get-Item '%s'
+Remove-Item $path -Recurse -Confirm:$false
+]]):format(trash_info.trash_file),
+    }, {
+      on_exit = function(_, code)
+        if code ~= 0 then
+          return cb("Error purging file")
+        end
+        cb()
+      end,
+    })
+    if jid <= 0 then
+      cb("Could not purge item")
+    end
+  elseif action.type == "move" then
+    local src_adapter = assert(config.get_adapter_by_scheme(action.src_url))
+    local dest_adapter = assert(config.get_adapter_by_scheme(action.dest_url))
+    if src_adapter.name == "files" then
+      local _, path = util.parse_url(action.src_url)
+      M.delete_to_trash(assert(path), cb)
+    elseif dest_adapter.name == "files" then
+      -- Restore
+      local _, dest_path = util.parse_url(action.dest_url)
+      assert(dest_path)
+      local entry = assert(cache.get_entry_by_url(action.src_url))
+      local meta = entry[FIELD_META] --[[@as {stat: uv_fs_t, trash_info: oil.TrashInfo, display_name: string}]]
+      local trash_info = meta.trash_info
+
+      local jid = vim.fn.jobstart({
+        "powershell",
+        "-NoProfile",
+        "-ExecutionPolicy",
+        "Bypass",
+        "-Command",
+        ([[
+$path = Get-Item '%s'
+$shell = New-Object -ComObject 'Shell.Application'
+$folder = $shell.NameSpace(0xa)
+$folder.ParseName($path.FullName).InvokeVerb('undelete')
+]]):format(trash_info.trash_file),
+      }, {
+        stdout_buffered = true,
+        on_exit = function(_, code)
+          if code ~= 0 then
+            return cb("Error restoring file")
+          end
+          cb()
+        end,
+      })
+      if jid <= 0 then
+        cb("Could not restore file")
+      end
+    end
+  elseif action.type == "copy" then
+    -- TODO: ...
```

Comment:
I was scared of using the luv functions directly with files from the recycle bin (or the oil utiliy functions), so wanted to try if using regular poweshell operations worked first. It seems like using luv (and oil utility functions) directly does work to move/rename/delete this files, so I'll change that to (the windows adapter will end up looking a lot like the freedesktop one xd).

Reply:
Also, I found [this random answer in stackoverflow](https://superuser.com/a/1736690) about how recycle bin works on Windows. Would it be preffered to handle this in a low level way like in the freedesktop implementation? Given the lack of proper Windows documentation, I'm inclined to let poweshell send files to the recycle bin, but use a implementation similar to the freedesktop one for moving, coping and purging files (I'm now realizing that this would mean not only handling the deleted file itself, but also the file with the metadata, just like on freedesktop)

Reply:
I just pushed my last changes doing exactly this, let me know your thoughts on the matter. Powershell is still needed for listing recycle bin files and for sending files to the trash bin